extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/util/ThriftUtils.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.data.input.thrift.util;
+
+import io.druid.java.util.common.logger.Logger;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TDeserializer;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TJSONProtocol;
+
+import java.nio.charset.StandardCharsets;
+
+import static java.util.Objects.requireNonNull;
+
+public final class ThriftUtils
+{
+
+  private static final Logger log = new Logger(ThriftUtils.class);
+
+  private static final ThreadLocal<TSerializer> SERIALIZER = new ThreadLocal<TSerializer>()
+  {
+    @Override
+    protected TSerializer initialValue()
+    {
+      return new TSerializer();
+    }
+  };
+
+
+  private static final ThreadLocal<TDeserializer> DESERIALIZER = new ThreadLocal<TDeserializer>()
+  {
+    @Override
+    protected TDeserializer initialValue()
+    {
+      return new TDeserializer();
+    }
+  };
+
+  private static final ThreadLocal<TDeserializer> DESERIALIZER_COMPACT = new ThreadLocal<TDeserializer>()
+  {
+    @Override
+    protected TDeserializer initialValue()
+    {
+      return new TDeserializer(new TCompactProtocol.Factory());
+    }
+  };
+
+  private static final ThreadLocal<TDeserializer> DESERIALIZER_JSON = new ThreadLocal<TDeserializer>()
+  {
+    @Override
+    protected TDeserializer initialValue()
+    {
+      return new TDeserializer(new TJSONProtocol.Factory());
+    }
+  };
+
+
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  public static byte[] decodeB64IfNeeded(final byte[] src)
+  {
+    if (requireNonNull(src).length <= 0) {
+      return EMPTY_BYTES;
+    }
+    final byte last = src[src.length - 1];
+    return (0 == last || '}' == last) ? src : Base64.decodeBase64(src);
+  }
+
+  /**
+   * Attempt to determine the protocol used to serialize some data.
+   * <p>
+   * The guess algorithm is copied from TProtocolUtil.java of thrift java runtime library.
```

Comment:
Can we attach a license file for this module that provides attribution for where this code is copied from?